src/lib/structures/modelMap.ml

```diff
@@ -30,100 +30,179 @@
+        match Value.Map.find_opt ret_val acc with
+        | Some fib ->
+          Value.Map.add ret_val (Fiber.add arg_vals fib) acc
+        | None ->
+          Value.Map.add ret_val (Fiber.of_list [arg_vals]) acc
+      ) graph Value.Map.empty
+
+  let pp_inverse ppf rel =
+    let rec aux ppf seq =
+      match seq () with
+      | Seq.Nil -> ()
+      | Cons ((ret_val, _), seq) when Stdcompat.Seq.is_empty seq ->
+        Fmt.pf ppf "%a" Value.pp ret_val
+      | Cons ((ret_val, fiber), seq) ->
+        Fmt.pf ppf "(@[<hv>ite %a@ %a@ %a)@]"
+          Fiber.pp fiber
+          Value.pp ret_val
+          aux seq
+    in
+    aux ppf (Value.Map.to_seq rel)
+end
 
-module V = Set.Make
+module P = Map.Make
     (struct
-      type t = (E.t * (X.r * string)) list * (X.r * string)
-      let compare (l1, (v1,_)) (l2, (v2,_)) =
-        let c = X.hash_cmp v1 v2 in
-        if c <> 0 then c
-        else
-          let c = ref 0 in
-          try
-            List.iter2
-              (fun (_,(x,_)) (_,(y,_)) ->
-                 let d = X.hash_cmp x y in
-                 if d <> 0 then begin c := d; raise Exit end
-              ) l1 l2;
-            !c
-          with
-          | Exit -> !c
-          | Invalid_argument _ -> List.length l1 - List.length l2
+      type t = sig_ [@@deriving ord]
     end)
 
-type key = P.key
-type elt = V.t
-
 type t = {
-  values : V.t P.t;
+  values : Graph.t P.t;
   suspicious : bool;
 }
 
-let is_suspicious_name hs =
-  match Hstring.view hs with
-  | "@/" | "@%" | "@*" -> true
-  | _ -> false
-
-(* The model generation is known to be imcomplete for FPA and Bitvector
-   theories. *)
-let is_suspicious_symbol = function
-  | Sy.Op (Float | Fixed | Abs_int | Abs_real | Sqrt_real
-          | Sqrt_real_default | Sqrt_real_excess
-          | Real_of_int | Int_floor | Int_ceil
-          | Max_int | Max_real | Min_int | Min_real
-          | Pow | Integer_log2 | Int2BV _ | BV2Nat
-          | BVand | BVor | Integer_round) -> true
-  | Sy.Name (hs, _, _) when is_suspicious_name hs -> true
-  | _ -> false
-
-let add ((sy, _, _) as p) v { values; suspicious } =
-  let prof_p = try P.find p values with Not_found -> V.empty in
-  let values =
-    if V.mem v prof_p then values
-    else P.add p (V.add v prof_p) values
-  in
-  { values; suspicious = suspicious || is_suspicious_symbol sy }
-
-let iter f { values; _ } =
-  P.iter (fun ((sy, _, _) as key) value ->
-      match sy with
-      | Sy.Name (_, _, true) ->
-        (* We don't print constants defined by the user. *)
-        ()
-      | _ -> f key value
-    ) values
-
-let[@inline always] is_suspicious { suspicious; _ } = suspicious
-
-let fold f { values; _ } acc = P.fold f values acc
-
-let empty = { values = P.empty; suspicious = false }
-
-let is_empty { values; _ } = P.is_empty values
+let add ((_, arg_tys, _) as sig_) arg_vals ret_val { values; suspicious } =
+  assert (List.compare_lengths arg_tys arg_vals = 0);
```

Comment:
Add some documentation about this assertion

Reply:
I replaced it by an `invalid_argument` as explained offline ;D